WIP - Add `reverseProxy` directive #16844

TobiasPfeifer opened the pull request:
Directives to enable proxying of single requests to an (internal) remote backend.
#16844

Commit:
Add `reverseProxy` directive #16844

Directives to enable proxying of single requests to an (internal) remote backend.

TobiasPfeifer commented:
This feature is not complete yet, but I'd like to have some feedback on my current work.

Commit:
remove 'Created by' header, preserve chronological order in Directives and some minor cleanup

johanandren commented:
(made it obvious it's not meant for reviewing)

ktoso commented:
Not meant for merging but for reviewing yes I thought :)

ktoso commented:
Btw related header feature https://tools.ietf.org/html/rfc7239#section-5.2

johanandren commented:
sorry, I thought merging but my fingers wrote reviewing

ktoso commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala:
```diff
@@ -0,0 +1,142 @@
+
+trait ReverseProxyDirectives {
+  type RequestInterceptor = (HttpRequest) ⇒ HttpRequest
+  type ResponseInterceptor = ((HttpResponse) ⇒ HttpResponse)
```
Not sure we need the interceptors.
Outgoing is simply a mapRequest in the directives, and incoming ireaponse is a map on a future or plain function application hmhm

ktoso commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala:
```diff
@@ -0,0 +1,142 @@
+    }
+  }
+
+  def forwardToUri(uri: Uri, circuitBreakerOption: Option[CircuitBreaker] = None)(implicit system: ActorSystem): Directive1[Try[HttpResponse]] = {
```
Aavoid requiring the actor system. Instead perhaps a confih object with the breaker and optional pool?

TobiasPfeifer replied:
Well, to make a request to another (remote) server, we either need an `ActorSystem` (implicit or explicit) to construct a `Function` `(HttpRequest) => Future[HttpResponse]` or a `Function` `(HttpRequest) => Future[HttpResponse]` directly.

If I get you right, you'd prefer to explicitly pass a `ReverseProxyConfig` to the Directives:
```
trait ReverseProxyConfig {
  val circuitBreakerOption: Option[CircuitBreaker]
  val requestExecutor: (HttpRequest) => Future[HttpResponse]
}
```

But how can we provide a default `requestExecutor`, if we don't have an `ActorSystem` in scope?
If the `requestExecutor` should be configurable via Typesafe Config, how can we construct a Function `(HttpRequest) => Future[HttpResponse]` without an `ActorSystem` in implicit scope?

ktoso replied:
If you rebase your branch now you'll notice a new directive: https://github.com/akka/akka/pull/21097 `extractActorSystem`.
So you can pull out the system out of a hat here, instead of passing it through 👍 

I'll think about the other options now

Commit:
remove Request/Reponse Interceptor  and introduce a ReverseProxyConfig

Request/Reponse Interceptor are simply mapRequest, mapResponse and mapRouteResult
By introducing an optional ReverseProxyConfig parameter we don't need to pass an ActorSystem through all directives to construct a RequestExecutor. If a ReverseProxyConfig is provided there is no need to have an ActorSystem in scope. Otherwise the ActorSystem is extracted to construct a RequestExecutor

TobiasPfeifer commented:
push.
How about the current scala API? I'm not sure if we need `complete...` `Directives` at all at this level.
I'd like to first finish the api before doing docs/examples.

ktoso commented:
Hi @TobiasPfeifer, I'm really sorry for not getting back to you on this one sooner.
We moved Akka HTTP into it's own repo which took some preparation.

So the ticket is now: https://github.com/akka/akka-http/issues/240
And the PR would be best to re-target at that repo.

cc @jrudolph please review if you have a moment :)

ktoso commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala:
```diff
@@ -23,59 +23,70 @@ import scala.concurrent.Future
   }
 }
 
-case object ProxyHeaderRequestInterceptor extends ((HttpRequest) ⇒ HttpRequest) {
```
good 👍

ktoso commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala:
```diff
@@ -140,3 +151,13 @@ case class CachedHostConnectionPoolRequestExecutor(implicit val system: ActorSys
   }
 }
 
+case class ReverseProxyConfig(
```
`final`

ktoso replied:
Perhaps make it not a case class? They're nearly impossible to evolve in a binary compatible way.

See here for inspiration https://github.com/akka/akka-http/blob/master/akka-http-core/src/main/scala/akka/http/javadsl/settings/ServerSettings.scala#L23

Hide it's fields behind an interface and create an `object ReverseProxyConfig` that serves as factory.
Impl can be a case class, but we need to hide it from users.

ktoso commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala:
```diff
@@ -23,59 +23,70 @@ import scala.concurrent.Future
 }
 
-case object ProxyHeaderRequestInterceptor extends ((HttpRequest) ⇒ HttpRequest) {
+private[akka] case object ProxyHeaderRequestInterceptor extends ((HttpRequest) ⇒ HttpRequest) {
```
please also add `/** INTERNAL API */` comment, as `private[akka]` compiles down to `public` for Java (in bytecode)

ktoso commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala:
```diff
@@ -23,59 +23,70 @@ import scala.concurrent.Future
+    }
+  }
+
+  private[akka] def getOrCreateDefaultConfig(configOption: Option[ReverseProxyConfig]): Directive1[ReverseProxyConfig] = {
```
can this be completely `private`? We're polluting public namespace currently

ktoso commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala:
```diff
@@ -23,59 +23,70 @@ import scala.concurrent.Future
+      case Some(config) ⇒ provide(config)
+      case None ⇒ {
+        extractMaterializer.flatMap { mat ⇒
+          extractActorSystem.flatMap { system ⇒
```
good 👍

ktoso commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala:
```diff
@@ -23,59 +23,70 @@ import scala.concurrent.Future
-          circuitBreakerOption.fold(onComplete(responseFuture))(circuitBreaker ⇒ onCompleteWithBreaker(circuitBreaker)(responseFuture))
+      extractRequest.flatMap { request ⇒
+        getOrCreateDefaultConfig(configOption).flatMap { reverseProxyConfig ⇒
+          val responseFuture: Future[HttpResponse] = reverseProxyConfig.requestExecutor.apply(request)
```
I'm unsure which executor the future will execute on here, do people have a way of configuring that?

ktoso commented on akka-http/src/main/scala/akka/http/scaladsl/server/Directives.scala:
```diff
@@ -36,7 +36,7 @@ trait Directives extends RouteConcatenation
   with WebSocketDirectives
-  with ReverseProxyDirectives
   with FramedEntityStreamingDirectives
+  with ReverseProxyDirectives
```
ok, this is not bin-compat but that's fine _currently_, and we'll deprecate the trait mixin style in the new repo (`trait Directives`) so this should be compatible from nowon anyway, good

ktoso commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala:
```diff
@@ -86,8 +83,8 @@ case object ProxyHeaderRequestInterceptor extends ((HttpRequest) ⇒ HttpRequest
 
-    var xRealIpHeaderOption = headers.collectFirst(xRealIpPF)
-    var xForwardedForHeaderOption = headers.collectFirst(xForwardedForPF)
+    val xRealIpHeaderOption = headers.collectFirst(xRealIpPF)
```
good

ktoso commented on akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectivesSpec.scala:
```diff
@@ -0,0 +1,92 @@
+  "The 'forwardRequest' directive" should {
+    "intercept the incoming HttpRequest" in {
+      val i: AtomicInteger = new AtomicInteger(0)
+      val route = forwardRequest(x ⇒ Future.successful(Ok.copy(headers = x.headers, entity = x.entity)), request ⇒ { i.incrementAndGet(); request.withEntity("executed") }) { completeTryResponse(_) }
```
nice

ktoso commented:
This looks very good, just ad a few questions and needs to be opened against the new repo: http://github.com/akka/akka-http
Otherwise would love to include it!

leachbj commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala:
```diff
@@ -0,0 +1,163 @@
+    extractExecutionContext.flatMap { implicit ec ⇒
+      extractRequest.flatMap { request ⇒
+        getOrCreateDefaultConfig(configOption).flatMap { reverseProxyConfig ⇒
+          val responseFuture: Future[HttpResponse] = reverseProxyConfig.requestExecutor.apply(request)
```
Because of the way akka-http times out, the HttpResponse won't be consumed if it takes too long.  Assuming you just want to use the existing timeouts;

    def processTimeout(proxyResult: Future[HttpResponse])(implicit ec: ExecutionContext, mat: Materializer): (HttpRequest) => HttpResponse = { req =>
      proxyResult.onSuccess {
        case proxyResponse => proxyResponse.entity.dataBytes.runWith(Sink.ignore)
      }

      HttpResponse(StatusCodes.ServiceUnavailable, entity =
        HttpEntity.Strict(ContentTypes.`application/json`,
          ByteString("""{"error":"The server was not able to produce a timely response to your request."}""")))
     }

     withRequestTimeoutResponse(processTimeout(responseFuture) {
       ...
     }

leachbj commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala:
```diff
@@ -0,0 +1,163 @@
+
+    val updatedXRealIpHeaderOption = ipOption.map(`X-Real-Ip`(_))
+    val updatedXForwardedForHeaderOption = remoteAddressOption.map(remoteAddresss ⇒ xForwardedForHeaderOption.fold(`X-Forwarded-For`(remoteAddresss))(h ⇒ `X-Forwarded-For`(h.addresses ++ Seq(remoteAddresss))))
+
```
you need to filter Connection and `Timeout-Access` headers (might be worth having a look over https://www.mnot.net/blog/2011/07/11/what_proxies_must_do).

leachbj commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala:
```diff
@@ -0,0 +1,163 @@
+private[akka] case object ProxyHeaderRequestInterceptor extends ((HttpRequest) ⇒ HttpRequest) {
+  override def apply(httpRequest: HttpRequest): HttpRequest = {
+    val headers: Seq[HttpHeader] = httpRequest.headers
+    val remoteAddressHeaderOption = headers.collectFirst({ case h: `Remote-Address` ⇒ h })
```
val remoteAddressHeaderOption = req.header[`Remote-Address`]

leachbj commented on akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala:
```diff
@@ -0,0 +1,163 @@
+  }
+}
+
+case class OutgoingRequestExecutor(implicit val system: ActorSystem, implicit val materializer: Materializer) extends ((HttpRequest) ⇒ Future[HttpResponse]) {
```
I don't think the next two really help, they'll basically just do the same as singleRequest since the flow is materialized for each request.

TobiasPfeifer replied:
will this have a negative impact on performance? Would you prefer to use `Flow[HttpRequest, HttpResponse, ???]` instead of `(HttpRequest) ⇒ Future[HttpResponse]`

ktoso replied:
It is mostly a question _when_ we materialize. 
Since the outgoing connection will need to materialize using that Flow (or function) anyway we won't gain performance from this I think. I like the idea of requesting a Flow[Request, Response, _] from users, since that's what they're already used to in any case.

TobiasPfeifer commented:
Thanks for the feedback so far. I applied the current state to akka/akka-http and created a new PR there: akka/akka-http#324
I'll continue my work there and therefore close this PR now.

ktoso commented:
Thanks a lot :)
